pkg/sql/sem/builtins/aggregate_builtins.go

```diff
@@ -477,10 +477,98 @@ const sizeOfJSONAggregate = int64(unsafe.Sizeof(jsonAggregate{}))
 const sizeOfBitAndAggregate = int64(unsafe.Sizeof(bitAndAggregate{}))
 const sizeOfBitOrAggregate = int64(unsafe.Sizeof(bitOrAggregate{}))
 
+// singleDatumAggregateBase is a utility struct that helps aggregate builtins
+// that store a single datum internally track their memory usage related to
+// that single datum.
+// It will reuse tree.EvalCtx.SingleDatumAggMemAccount when non-nil and will
+// *not* close that account upon its closure; if it is nil, then a new memory
+// account will be created specifically for this struct and that account will
+// be closed upon this struct's closure.
+type singleDatumAggregateBase struct {
+	mode singleDatumAggregateBaseMode
+	acc  *mon.BoundAccount
+	// accountedFor indicates how much memory (in bytes) have been registered
+	// with acc.
+	accountedFor int64
+}
+
+// singleDatumAggregateBaseMode indicates the mode in which
+// singleDatumAggregateBase operates with regards to resetting and closing
+// behaviors.
+type singleDatumAggregateBaseMode int
+
+const (
+	// sharedSingleDatumAggregateBaseMode is a mode in which the memory account
+	// will be grown and shrunk according the corresponding aggregate builtin's
+	// memory usage, but the account will never be cleared or closed. In this
+	// mode, singleDatumAggregateBaseMode is *not* responsible for closing the
+	// memory account.
+	sharedSingleDatumAggregateBaseMode singleDatumAggregateBaseMode = iota
+	// nonSharedSingleDatumAggregateBaseMode is a mode in which the memory
+	// account is "owned" by singleDatumAggregateBase, so the account can be
+	// cleared and closed by it. In fact, singleDatumAggregateBase is
+	// responsible for the account's closure.
+	nonSharedSingleDatumAggregateBaseMode
+)
+
+// makeSingleDatumAggregateBase makes a new singleDatumAggregateBase. If
+// evalCtx has non-nil SingleDatumAggMemAccount field, then that memory account
+// will be used by the new struct which will operate in "shared" mode
+func makeSingleDatumAggregateBase(evalCtx *tree.EvalContext) singleDatumAggregateBase {
+	if evalCtx.SingleDatumAggMemAccount == nil {
+		newAcc := evalCtx.Mon.MakeBoundAccount()
+		return singleDatumAggregateBase{
+			mode: nonSharedSingleDatumAggregateBaseMode,
+			acc:  &newAcc,
+		}
+	}
+	return singleDatumAggregateBase{
+		mode: sharedSingleDatumAggregateBaseMode,
+		acc:  evalCtx.SingleDatumAggMemAccount,
+	}
+}
+
+// updateMemoryUsage updates the memory account to reflect the new memory
+// usage. If any memory has been previously registered with this struct, then
+// the account is updated only by the delta between previous and new usages,
+// otherwise, it is grown by newUsage.
+func (b singleDatumAggregateBase) updateMemoryUsage(ctx context.Context, newUsage int64) error {
+	if err := b.acc.Grow(ctx, newUsage-b.accountedFor); err != nil {
+		return err
+	}
+	b.accountedFor = newUsage
+	return nil
+}
+
+func (b singleDatumAggregateBase) reset(ctx context.Context) {
```

Comment:
I think the receivers in these methods need to be pointers otherwise `accountedFor` won't be updated correctly.